sw/device/lib/dif/dif_uart.c

```diff
@@ -0,0 +1,409 @@
+                           UART_FIFO_CTRL_TXILVL_OFFSET);
+
+  return true;
+}
+
+bool dif_uart_bytes_send(const dif_uart_t *uart, const uint8_t *data,
+                         size_t bytes_requested, size_t *bytes_written) {
+  if (uart == NULL || bytes_written == NULL) {
+    return false;
+  }
+
+  // |bytes_written| is an optional parameter
+  if (bytes_written == NULL) {
+    (void)uart_bytes_send(uart, data, bytes_requested);
+  } else {
+    *bytes_written = uart_bytes_send(uart, data, bytes_requested);
+  }
+
+  return true;
+}
+
+bool dif_uart_bytes_receive(const dif_uart_t *uart, size_t bytes_requested,
+                            uint8_t *data, size_t *bytes_read) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // |bytes_read| is an optional parameter
+  if (bytes_read == NULL) {
+    (void)uart_bytes_receive(uart, bytes_requested, data);
+  } else {
+    *bytes_read = uart_bytes_receive(uart, bytes_requested, data);
+  }
+
+  return true;
+}
+
+bool dif_uart_byte_send_polled(const dif_uart_t *uart, uint8_t byte) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // Busy wait for the TX FIFO to free up
+  while (uart_tx_full(uart)) {
+  }
+
+  (void)uart_bytes_send(uart, &byte, 1);
+
+  // Busy wait for the TX FIFO to be drained (the byte has been processed by
+  // the UART hardware).
+  while (!uart_tx_empty(uart)) {
+  }
+
+  return true;
+}
+
+bool dif_uart_byte_receive_polled(const dif_uart_t *uart, uint8_t *byte) {
+  if (uart == NULL || byte == NULL) {
+    return false;
+  }
+
+  // Busy wait for the RX message in the FIFO
+  while (uart_rx_empty(uart)) {
+  }
+
+  (void)uart_bytes_receive(uart, 1, byte);
+
+  return true;
+}
+
+bool dif_uart_irq_state_get(const dif_uart_t *uart,
+                            dif_uart_interrupt_t irq_type,
+                            dif_uart_enable_t *state) {
+  if (uart == NULL || state == NULL) {
+    return false;
+  }
+
+  // Get the requested interrupt state (enabled/disabled)
+  ptrdiff_t offset = uart_irq_offset_get(irq_type);
+  bool enabled = reg32_get_bit(uart->base_addr, UART_INTR_STATE(0), offset);
+  if (enabled) {
+    *state = kDifUartEnable;
+  } else {
+    *state = kDifUartDisable;
+  }
+
+  return true;
+}
+
+bool dif_uart_irqs_disable(const dif_uart_t *uart, uint32_t *state) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // Get the current interrupts state
+  uint32_t reg = reg32_read(uart->base_addr, UART_INTR_ENABLE(0));
+
+  // Disable all UART interrupts
+  reg32_write(uart->base_addr, UART_INTR_ENABLE(0), 0u);
+
+  // Pass the previous interrupt state to the caller
+  if (state != NULL) {
+    *state = reg;
+  }
+
+  return true;
+}
+
+bool dif_uart_irqs_restore(const dif_uart_t *uart, uint32_t state) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // Restore the interrupt state
+  reg32_write(uart->base_addr, UART_INTR_ENABLE(0), state);
+
+  return true;
+}
+
+bool dif_uart_irq_enable(const dif_uart_t *uart, dif_uart_enable_t enable,
```

Comment:
`dif_uart_irq_state_get` has `irq_type` after `uart` as the second argument. It would be great if this function did the same.

Reply:
done